sys-fs/mergerfs: new package, version 2.24.2

ahippo opened the pull request:
Add sys-fs/mergerfs package -- IMHO, a superior alternative to mhddfs.

For now, I put myself as the package maintainer.
I'll be glad to maintain the package although I'm not a Gentoo developer (yet).

Bug: https://bugs.gentoo.org/587956
Package-Manager: Portage-2.3.40, Repoman-2.3.9

Another take on mergerfs ebuild.
It's not based in [mergerfs-2.13.1.ebuild](https://bugs.gentoo.org/attachment.cgi?id=439650) and is for much newer version.

ahippo commented:
`build.log` with `USE=-xattr` is [attached](https://bugs.gentoo.org/attachment.cgi?id=543130) to Gentoo bug 587956.
`build.log` with `USE=xattr` is [also attached](https://bugs.gentoo.org/attachment.cgi?id=543132) to the same bug.

ahippo commented:
Refreshed the mergerfs LDFLAGS patch with the one accepted upstream: the only differences are the commit hash and message.

mgorny commented on sys-fs/mergerfs/mergerfs-2.24.2.ebuild:
```diff
@@ -0,0 +1,98 @@
+
+DESCRIPTION="A featureful FUSE union filesystem across multiple drives"
+HOMEPAGE="https://github.com/trapexit/mergerfs"
+# tarball under download/ has prebuilt ChangeLog and AUTHORS,
```
Dedicated release tarballs are always preferable over snapshots, so there's no need to explain it in a comment.

ahippo replied:
Well, release tags can be verified against the author's GPG signature, while (in this particular case) release tarballs can not.
Also, in this case, the release tarball contains pre-compiled binaries, which is not always desired.
That's why I tried to explain, what's particularly useful in the release tarball.

I can surely remove the comment per your suggestion.

mgorny replied:
But autogenerated github tarballs don't contain enough data to verify a signature either, and we can't use git directly. Prebuilt files aren't much of an issue as long as they end up being recompiled.

ahippo replied:
In fully automated way -- no.
But manually, I could have cloned the repo, verified the tag signature, generated a tarball from the tag, and compared it to the one obtained from github; and then set the right checksum in `Manifest` file.

Anyway, I'll remove the comment.

mgorny commented on sys-fs/mergerfs/mergerfs-2.24.2.ebuild:
```diff
@@ -0,0 +1,98 @@
+)
+
+domake() {
+	local makeargs=( "$@" )
```
Do you really need two vars here? Either include `"${@}"` in `makevars`, or pass it directly to `emake`.

ahippo replied:
OK, will pass it to `emake`.

mgorny commented on sys-fs/mergerfs/mergerfs-2.24.2.ebuild:
```diff
@@ -0,0 +1,98 @@
+	local makeargs=( "$@" )
+	local makevars=(
+		GIT=''
+		MKDIR="$(type -P mkdir)"
```
Are all those overrides (up to `INSTALL`) really necessary?

ahippo replied:
Mostly yes.
I've double-checked and will drop a couple of them.
* `GIT` -- will drop
* `MKDIR` -- used during `make all` and `make install`
* `TOUCH` -- used during `make all`
* `CP` -- used during `make install`
* `RM` -- will drop
* `LN` -- used during `make install`
* `INSTALL` -- used during `make install`

mgorny replied:
I've just looked at the Makefile and it's so horrible I can't find appropriate words to describe it. You need to DEPEND on `sys-apps/which` btw. Not that using `which` makes any sense (or `type -P` either, why would anyone need full paths there?!).

ahippo replied:
No, I override these variables exactly to avoid `BDEPEND` on `which`.
Yeah, fair enough, I could have omitted `type -P` and used the commands without full paths.

mgorny commented on sys-fs/mergerfs/mergerfs-2.24.2.ebuild:
```diff
@@ -0,0 +1,98 @@
+	default
+
+	# don't rebuild version.hpp
+	cat > tools/update-version <<EOFF
```
`|| die`

Also you can use `<<-EOF` to strip whitespace and allow indenting stuff, e.g.:

```
  cat > foo <<-EOF || die
    bar baz
  EOF

ahippo replied:
Oh, good catch!
Forgot about `|| die`.
(so used to `set -e`)

Will convert to the indented here-doc too.

mgorny commented on sys-fs/mergerfs/mergerfs-2.24.2.ebuild:
```diff
@@ -0,0 +1,98 @@
+exit 0
+EOFF
+
+	pushd libfuse >/dev/null
```
I'm confused. Why do all actions except for install need to be done in `libfuse`?

Also, all external commands need `||die`.

ahippo replied:
mergerfs bundles its own copy of (patched) libfuse.
(it does compile against vanilla libfuse-2.7.9, but I don't know the effect of effectively dropping those patches)

Thus, it tries to `autoreconf`, `configure`, and `make` it during `make all`.
First, it somewhat violates `src_prepare()/src_configure()/src_compile()` separation.
But most importantly, it likely ignores all toolchain-related variables, {CXX,LD}FLAGS, make opts, which are normally passed in by Portage.
So, I replaced `autoreconf` with explicit `eautoreconf` in the ebuild, `configure` with `econf`, `make` with `emake` to make sure that everything is sane.

There is nothing related to `libfuse` in `src_install()` because it's compiled as a static library,
and thus linked in statically into the resulting executable.

mgorny replied:
Hmm, but now mergerfs is built in `src_install` instead of `src_compile`, isn't it?

ahippo replied:
No, it's built in the end of `src_compile()` by calling `domake`.

mgorny commented on sys-fs/mergerfs/mergerfs-2.24.2.ebuild:
```diff
@@ -0,0 +1,98 @@
+
+EAPI=7
+
+inherit toolchain-funcs autotools
```
Sort eclasses lexically.

ahippo replied:
OK.

mgorny commented on sys-fs/mergerfs/metadata.xml:
```diff
@@ -0,0 +1,16 @@
+<pkgmetadata>
+	<maintainer type="person">
+		<email>ahippo@yandex.ru</email>
+	</maintainer>
```
Who's gonna commit for you?

ahippo replied:
I'll add proxy-maint as with the other PR.
Somehow missed this in the PR/Proxy_Maint instructions.

ahippo replied:
(becoming a Gentoo developer would be cool too, but I hope the PR will get merged sooner than that)

ahippo commented:
I've pushed the updated changeset, which should address @mgorny's comments:
* sorted `inherit`ed eclasses;
* removed comment regarding `SRC_URI`;
* removed `DOCS` variable (since the default does the same);
* dropped `makeargs` variable and pass `$@` directly to `emake`;
* removed `GIT` and `RM` overrides;
* added `|| die` for all non-EAPI commands;
* indented the here-doc used to overwrite `tools/update-version`;
* added Proxy Maintainers to the list of maintainers;
* converted tabs to spaces in metadata.xml (I checked a few metadata.xmls in the tree, and the style appears to be inconsistent; devmanual permits any style too);
* rebased on top of the current master.

mgorny commented:
I'm sorry but the deeper we get into this, the more I feel that we really don't want this package in Gentoo. I'm sorry that I've wasted your time on improving the ebuild but without major changes upstream, it's a no-go. In particular, the patched libfuse needs to stop being a requirement, and Makefile needs to be seriously sanitized. Feel free to resubmit once that happens.

ahippo commented:
> I'm sorry but the deeper we get into this, the more I feel that we really don't want this package in Gentoo. I'm sorry that I've wasted your time on improving the ebuild but without major changes upstream, it's a no-go. In particular, the patched libfuse needs to stop being a requirement, and Makefile needs to be seriously sanitized. Feel free to resubmit once that happens.

Well, it does compile (and run) against system libfuse-2.9.7.
But I'll check with upstream if there would be any issues.
One apparent difference is that it won't be possible to specify number of FUSE threads to use.
(like in any other FUSE-based fs)

Regarding the Makefile, yeah, it's hand written, calls autotools on libfuse, and a few other things
However, the ebuild covers all of its deficiencies and produces properly compiled binaries.

Could you, please, be more specific about what you want to see improved regarding the Makefile?
I could possibly convert it to CMake, but without a specific set of goal, it's hard to see if I'm on track.

mgorny commented:
If you could really verify *what* is changed regarding libfuse, so we'd know if it's sane to use, I suppose that would make things easier.

As for Makefile, it's just plain horrible pile of hacks:
1. You shouldn't really use absolute paths (gotten from `which` or otherwise), shell evaluates PATH for a reason.
2. Whenever that is done to determine whether something is available, there's a certainly a better way of doing the same thing.
3. Determining support for xattr forces static path to includedir which is plain broken for Prefix.
4. CFLAGS forces `-g` which should be user-configurable (as part of optimization-debug control).
5. CPPFLAGS are not respected.
6. There's no way to skip bundled libfuse.
7. Also nested make calls should be prefixed with `+` and call `$(MAKE)` to pass job control over (not that we want them).

I get that you can patch or override most of it. But the point is, we really don't have to include half of Makefile size of patches and overrides. If you could get some improvements upstream, that would be greatly preferable.

If you have any questions or want me to review some changes you've prepared for upstream, feel free to ping me here, on IRC (`mgorny`) or via mail (`mgorny@g.o`).

ahippo commented:
> If you could really verify _what_ is changed regarding libfuse, so we'd know if it's sane to use, I suppose that would make things easier.

As far as I can tell, there are changes to the way threads are created.
Instead of creating them as requests arrive, the threads are pre-created in advance.
Also, it's possible to specify the number of threads as a command-line option.
The changes also reduce locking contention while handling requests.

I checked with upstream, and he doesn't seem to see big problems with system libfuse.
At least, it compiles fine with `=sys-fs/fuse-2.9.7`.
Anyway, I'll send a PR upstream and see how it goes.

> 1. You shouldn't really use absolute paths (gotten from which or otherwise), shell evaluates PATH for a reason.

Makes sense, should be a trivial change.

> 2. Whenever that is done to determine whether something is available, there's a certainly a better way of doing the same thing.

Ok, will try to run corresponding utilities to see if they're available.

> 3. Determining support for xattr forces static path to includedir which is plain broken for Prefix.

Not sure what to do here.
xattr doesn't ship `.pc` file.
Add `WITH_XATTR` kind of variable (analogous to `configure --with-xattr=`) to point to the actual include path?

`XATTR_AVAILABLE` variable would need to be overridden in the ebuild anyway to avoid automagic dependencies.
(in case of `USE=-xattr`)

> 4. CFLAGS forces -g which should be user-configurable (as part of optimization-debug control).
> 5. CPPFLAGS are not respected.

OK, should be a trivial change.

> 6. There's no way to skip bundled libfuse.

This is almost ready for submission upstream: I added a new `INTERNAL_FUSE` variable to control if system-wide libfuse is picked up via `pkg-config`, or the bundled one is used.

> 7. Also nested make calls should be prefixed with + and call $(MAKE) to pass job control over (not that we want them).

Good catch!
Should be a trivial change.
I was still planning on `emake`ing subdirs in the ebuild explicitly in case of `INTERNAL_FUSE=1`/`USE=-external-fuse`.

> I get that you can patch or override most of it. But the point is, we really don't have to include half of Makefile size of patches and overrides. If you could get some improvements upstream, that would be greatly preferable.
> 
> If you have any questions or want me to review some changes you've prepared for upstream, feel free to ping me here, on IRC (mgorny) or via mail (mgorny@g.o).

Thank you for willing to help!
Would including you in the upstream PR discussion be an option?
(so that the same PR can be looked at by you and upstream at the same time)

mgorny commented:
> Would including you in the upstream PR discussion be an option?

Sure.

ahippo commented:
>> Would including you in the upstream PR discussion be an option?
> 
> Sure.

Awesome, thank you!
